op-test-sequencer/sequencer/backend/work/builders/fakepos/job.go

```diff
@@ -0,0 +1,253 @@
+package fakepos
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	opeth "github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/testutils"
+	"github.com/ethereum-optimism/optimism/op-test-sequencer/sequencer/backend/work"
+	"github.com/ethereum-optimism/optimism/op-test-sequencer/sequencer/seqtypes"
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type FakePoSEnvelope struct {
+	engine.ExecutionPayloadEnvelope
+}
+
+func (e *FakePoSEnvelope) ID() eth.BlockID {
+	return eth.BlockID{Hash: e.ExecutionPayload.BlockHash, Number: e.ExecutionPayload.Number}
+}
+
+func (e *FakePoSEnvelope) String() string {
+	return e.ID().String()
+}
+
+type Job struct {
+	id     seqtypes.BuildJobID
+	mu     sync.Mutex
+	logger log.Logger
+
+	b *Builder
+
+	parent   common.Hash
+	envelope engine.ExecutionPayloadEnvelope
+
+	head      *types.Header
+	safe      *types.Header
+	finalized *types.Header
+
+	parentBeaconBlockRoot common.Hash
+}
+
+func (j *Job) ID() seqtypes.BuildJobID {
+	return j.id
+}
+
+func (j *Job) Cancel(ctx context.Context) error {
+	return nil
+}
+
+func (j *Job) setHeadSafeAndFinalized() {
+	j.head = j.b.geth.Backend.BlockChain().CurrentBlock() // default head
+	if j.parent != (common.Hash{}) {
+		j.head = j.b.geth.Backend.BlockChain().GetHeaderByHash(j.parent) // override head if parent is set
+	}
+
+	j.finalized = j.b.geth.Backend.BlockChain().CurrentFinalBlock()
+	if j.finalized == nil { // fallback to genesis if nothing is finalized
+		j.finalized = j.b.geth.Backend.BlockChain().Genesis().Header()
+	}
+	j.safe = j.b.geth.Backend.BlockChain().CurrentSafeBlock()
+	if j.safe == nil { // fallback to finalized if nothing is safe
+		j.safe = j.finalized
+	}
+
+	if j.head.Number.Uint64() > j.b.finalizedDistance { // progress finalized block, if we can
+		j.finalized = j.b.geth.Backend.BlockChain().GetHeaderByNumber(j.head.Number.Uint64() - j.b.finalizedDistance)
+	}
+	if j.head.Number.Uint64() > j.b.safeDistance { // progress safe block, if we can
+		j.safe = j.b.geth.Backend.BlockChain().GetHeaderByNumber(j.head.Number.Uint64() - j.b.safeDistance)
+	}
+
+	j.parentBeaconBlockRoot = fakeBeaconBlockRoot(j.head.Time) // parent beacon block root
+}
+
+func (j *Job) Open(ctx context.Context) error {
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	j.logger.Info("Open job", "id", j.id)
+
+	j.setHeadSafeAndFinalized()
+
+	var withdrawals []*types.Withdrawal
+	envelope, ok := j.b.envelopes[j.head.Hash()]
```

Comment:
Just curious: why do we need to cache the envelope we built earlier? Would it be possible to make the `fakepos Builder` more stateless and rely more on the `Job` instead?

Reply:
We cache the envelope needed, because it is needed if we want to replace a block - I couldn't replace a block without keeping/caching the original block and updating a few of its values.

You might be right that we don't need to strictly save all blocks and just build new ones, but this turned out to be difficult.